### PR TITLE
fix(*): propagate context for metrics aggregate

### LINKED ISSUE #4608: Kuma "metrics hijacking"/exposing application metrics times out often.
### What happened?

Wat happen?
------------------

When using the kuma metrics hijacking feature introduced in 1.7.0 to gather metrics from kafka exporter (which takes a long time since it forwards the requests to the actual kafka brokers) the kuma sidecar regularly times out. Prometheus allows you to set a max scrape timeout for this reason but according to the [docs](https://kuma.io/docs/1.7.x/policies/traffic-metrics/#expose-metrics-from-data-plane-proxies) kuma does not.

How reproduce?
----------------------

Install kafka (using the bitnami helmchart) and enable metrics exporter and hijacking, look at the kuma sidecar logs to attain deeper knowledge.

### PR DESCRIPTION
**Pull request by lukidzi**
### Summary

Instead of setting a timeout, we rely on context. When the client interrupts the request we are going to cancel all pending requests.

### Full changelog

* [Removed timeout]
* [Added context propagation]

### Issues resolved

Fix #4608 

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### Testing

- [ ] Unit tests
- [X] E2E tests
- [x] Manual testing on Universal
- [ ] Manual testing on Kubernetes

### Backwards compatibility

- [ ] Update [`UPGRADE.md`](/UPGRADE.md) with any steps users will need to take when upgrading.
- [ ] Add `backport-to-stable` label if the code follows our [backporting policy](/CONTRIBUTING.md#backporting)

### CODE REVIEW

File: app/kuma-dp/pkg/dataplane/metrics/server.go [hunk 1]
```diff
@@ -120,32 +119,43 @@ func rewriteMetricsURL(path string, port uint32, queryModifier QueryParametersMo
...
+	case <-ctx.Done():
+		return
+	case <-done:
+		for resp := range out {
+			if _, err := writer.Write(resp); err != nil {
+				logger.Error(err, "error while writing the response")
+			}
+			if _, err := writer.Write([]byte("\n")); err != nil {
+				logger.Error(err, "error while writing the response")
+			}
```
Comment (lahabana) [inline app/kuma-dp/pkg/dataplane/metrics/server.go:150]: Have we considered using: https://pkg.go.dev/golang.org/x/sync@v0.0.0-20220601150217-0de741cfad7f/errgroup ?

File: app/kuma-dp/pkg/dataplane/metrics/server.go [hunk 2]
```diff
@@ -120,32 +119,43 @@ func rewriteMetricsURL(path string, port uint32, queryModifier QueryParametersMo
...
 		}
 	}
 }
 
-func (s *Hijacker) getStats(url *url.URL, app ApplicationToScrape) []byte {
-	resp, err := s.httpClient.Get(rewriteMetricsURL(app.Path, app.Port, app.QueryModifier, url))
+func (s *Hijacker) getStats(url *url.URL, app ApplicationToScrape, context context.Context) []byte {
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/metrics/server.go:155]: nit: context is always the first argument and it's called `ctx` by the convention.

File: app/kuma-dp/pkg/dataplane/metrics/server.go [hunk 3]
```diff
@@ -120,32 +119,43 @@ func rewriteMetricsURL(path string, port uint32, queryModifier QueryParametersMo
...
+	req, _ := http.NewRequest("GET", rewriteMetricsURL(app.Path, app.Port, app.QueryModifier, url), nil)
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/metrics/server.go:156]: why do we ignore an error?

File: app/kuma-dp/pkg/dataplane/metrics/server.go (hunk 1 above)
Comment (lukidzi) [reply app/kuma-dp/pkg/dataplane/metrics/server.go:150]: I am not sure if that is the case here. We don't want to return an error just log it.

**Review by jakubdyszkiewicz — APPROVED** [review]

File: app/kuma-dp/pkg/dataplane/metrics/server.go (hunk 1 above)
Comment (lahabana) [reply app/kuma-dp/pkg/dataplane/metrics/server.go:150]: But the api enables you to do scatter gather using the context like you are doing here no?

Comment (lahabana) [reply app/kuma-dp/pkg/dataplane/metrics/server.go:150]: I guess maybe not because you want partial success to still go forward?

Comment (lukidzi) [reply app/kuma-dp/pkg/dataplane/metrics/server.go:150]: Yes, that is the case. If we get only one response we just want to log the failure and return the successful one.

**Pull request merged**